### PR TITLE
Added logic for journey forking.

### PR DESCRIPTION
**Pull request by joefitter**
* Extended Form.prototype.getNextStep method to check if form should fork
* Checks if `forks` array is provided, processes each fork if so
* Fork condition accepts a function (passed req, res) and should return truthy/falsy to determine if fork should be followed
* Fork condition also accepts a `{field, value}` map, this is checked against req.form.values and follows fork if condition is satisfied
* if forks are not provided, original functionality is maintained
* Added docs from hof-controllers, added relevant tests

### CODE REVIEW

**Comment by joefitter** [comment]
@JoeChapman @easternbloc @gavboulton this is the current hof logic for forking just moved up to this level including tests. If this PR is merged I will submit another one for the unreachable step invalidation logic, and further down the line it we could perhaps swap out forking for @daniel-ac-martin's earlier suggestion of extending the `next` property to accept a function.

File: README.md [hunk 1]
```diff
@@ -57,3 +57,68 @@ These methods are synchronous and take only the request and response obejct as a
...
+In this example, if the last condition resolves to true - even if the others also resolve to true - then the page at '/fork-page-three' will be loaded. The last condition to be met is always the fork used to determine the next step.
+
+```js
+
+'/my-page': {
+  next: '/next-page',
+  forks: [{
+    target: '/fork-page-one',
+    condition: function (req, res) {
+      return req.form.values['name'].length > 30;
```
Comment (gavboulton) [inline README.md]: Could we also use validator strings/objects here?

This is similar to how we direct users to exit pages (when a user doesn't meet a criterion of our service). In a field's validate array we have something like `{ type: 'equal', arguments: true, redirect: '/exit-page }`, if the field value isn't true the user is redirected to /exit-page.

File: test/spec/spec.form.js [hunk 2]
```diff
@@ -515,14 +514,162 @@ describe('Form Controller', function () {
...
             req.baseUrl = '/base';
-            form.successHandler(req, res);
-            res.redirect.should.have.been.calledWith('/base/success');
+            form.getNextStep(req, res).should.be.equal('/base/next-page');
+        });
+
+        describe('with a fork', function () {
+
+            describe('when the condition config is met', function () {
+                it('the next step is the fork target', function () {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: Do we need a `describe` around one `it`? Reading 'it the next step is the fork target' doesn't make sense.

File: test/spec/spec.form.js [hunk 3]
```diff
@@ -515,14 +514,162 @@ describe('Form Controller', function () {
...
+                        }
+                    }];
+                });
+
+                describe('and each condition is met', function () {
+                    beforeEach(function () {
+                        req.form = {values: {
+                            'example-radio': 'conditionMet',
+                            'example-email': 'conditionMet'
+                        }};
```
Comment (gavboulton) [inline test/spec/spec.form.js]: Could you explicitly set `req.form.values` in the `it` block here?

File: test/spec/spec.form.js (hunk 2 above)
Comment (joefitter) [inline test/spec/spec.form.js]: good point - I just copied this straight over from hof-controllers

File: test/spec/spec.form.js [hunk 4]
```diff
@@ -515,14 +514,162 @@ describe('Form Controller', function () {
...
         });
 
         it('prefixes redirect url with req.baseUrl', function () {
             req.baseUrl = '/base';
-            form.successHandler(req, res);
-            res.redirect.should.have.been.calledWith('/base/success');
+            form.getNextStep(req, res).should.be.equal('/base/next-page');
+        });
+
+        describe('with a fork', function () {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: 'forking journeys'?

File: test/spec/spec.form.js (hunk 3 above)
Comment (joefitter) [inline test/spec/spec.form.js]: and below

File: README.md (hunk 1 above)
Comment (joefitter) [inline README.md]: Hmm I guess so - I wasn't aware of the `redirect` property in validators - looks like they are both achieving the same result, with redirect acting pretty much the same as a fork

Comment (joefitter) [inline README.md]: @gavboulton do you think we should be using forks in the place of the field validate redirect?

Comment (gavboulton) [inline README.md]: @joefitter I think that makes more sense, that way all routing config is in the same place. We can recommend it as the preferred option but it would be good to keep the existing redirect for backwards compatibility, at least for now.

**Comment by joefitter** [comment]
@gavboulton @JoeChapman @easternbloc @daniel-ac-martin 👓 👀  pls :)

File: lib/form.js [hunk 5]
```diff
@@ -160,12 +160,25 @@ _.extend(Form.prototype, {
...
         var next = this.options.next || req.path;
         if (req.baseUrl !== '/') {
             next = req.baseUrl + next;
         }
-        return next;
+        var forks = this.options.forks || [];
+
+        // If a fork condition is met, its target supercedes the next property
+        return _.reduce(forks, function forkFold(result, value) {
+            var evalCondition = function evalCondition(condition) {
```
Comment (JoeChapman) [inline lib/form.js]: is it necessary to recreate this function for each iteration?

Comment (joefitter) [inline lib/form.js]: No, good point - this is how it ended up in hof-controllers but the logic is simpler now - appending /edit and redirecting to /confirm still live within hof.

I'll tidy this

**Comment by joefitter** [comment]
@JoeChapman @gavboulton @easternbloc @daniel-ac-martin are we happy for this to be merged?

**Comment by JoeChapman** [comment]
LGTM, @daniel-ac-martin @easternbloc @gavboulton ?

**Comment by gavboulton** [comment]
I'll take another look this afternoon.

File: test/spec/spec.form.js [hunk 6]
```diff
@@ -515,14 +514,144 @@ describe('Form Controller', function () {
...
         it('prefixes redirect url with req.baseUrl', function () {
             req.baseUrl = '/base';
-            form.successHandler(req, res);
-            res.redirect.should.have.been.calledWith('/base/success');
+            form.getNextStep(req, res).should.be.equal('/base/next-page');
+        });
+
+        describe('forking journeys', function () {
+
+            it('return the fork target if the condition config is met', function () {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: return to returns.

**Comment by joefitter** [comment]
@gavboulton if this PR can be merged I will another one ready to follow it up - https://github.com/UKHomeOffice/passports-form-controller/commit/11ef15fadd6473444003ba9efe0b4125ca279391 - this exposes the result of the fork conditional for use in a wizard level middleware - https://github.com/UKHomeOffice/passports-form-wizard/commit/5db203a280ec61cf979f198d6b9ee171577eb8bc

File: README.md (hunk 1 above)
Comment (gavboulton) [inline README.md]: @joefitter Did you look at using validators here?

File: test/spec/spec.form.js [hunk 7]
```diff
@@ -515,14 +514,144 @@ describe('Form Controller', function () {
...
+                    condition: function () {
+                        return false;
+                    }
+                }];
+                form.getNextStep(req, {}).should.equal('/next-page');
+            });
+
+        });
+
+        describe('with more than one fork', function () {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: This would make more sense under the 'forking journeys' describe.

File: test/spec/spec.form.js [hunk 8]
```diff
@@ -515,14 +514,144 @@ describe('Form Controller', function () {
...
+
+            describe('when the fields are the same', function () {
+
+                beforeEach(function () {
+                    req.form = {values: {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: Space before values.

File: test/spec/spec.form.js [hunk 9]
```diff
@@ -515,14 +514,144 @@ describe('Form Controller', function () {
...
+                        target: '/target-page-2',
+                        condition: {
+                            field: 'example-email',
+                            value: 'conditionMet'
+                        }
+                    }];
+                });
+
+                it('returns the last forks\' target if each condition is met', function () {
+                    req.form = {values: {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: Space before values.

File: test/spec/spec.form.js [hunk 10]
```diff
@@ -515,14 +514,144 @@ describe('Form Controller', function () {
...
+                        'example-radio': 'conditionMet',
+                        'example-email': 'conditionMet'
+                    }};
+                    form.getNextStep(req, {}).should.contain('/target-page-2');
+                });
+
+                it('returns the first forks\' target if the first condition is met', function () {
```
Comment (gavboulton) [inline test/spec/spec.form.js]: Do we need this test? It seems a bit misleading, like the behaviour is different. In this case the first fork's target is the first in the array but is still the last match.

File: README.md (hunk 1 above)
Comment (joefitter) [inline README.md]: If we look at using validators it should be a separate PR IMO - I would prefer to just move current functionality in the first instance and then iterate - @JoeChapman @easternbloc @daniel-ac-martin thoughts?

File: test/spec/spec.form.js (hunk 8 above)
Comment (joefitter) [inline test/spec/spec.form.js]: a linter would pick this up - I'll add an issue for adding/configuring linting

File: test/spec/spec.form.js (hunk 10 above)
Comment (joefitter) [inline test/spec/spec.form.js]: true - will remove

**Comment by joefitter** [comment]
@gavboulton @JoeChapman @easternbloc @daniel-ac-martin how are we feeling about this PR?

File: lib/form.js [hunk 11]
```diff
@@ -160,12 +160,25 @@ _.extend(Form.prototype, {
...
-    getNextStep: function (req) {
+    getNextStep: function (req, res) {
         var next = this.options.next || req.path;
         if (req.baseUrl !== '/') {
             next = req.baseUrl + next;
         }
-        return next;
+        var forks = this.options.forks || [];
+
+        var evalCondition = function evalCondition(condition) {
```
Comment (JoeChapman) [inline lib/form.js]: just being pedantic, could this just be a named function declaration rather than a named literal?

**Comment by JoeChapman** [comment]
LGTM

File: README.md [hunk 12]
```diff
@@ -57,3 +57,68 @@ These methods are synchronous and take only the request and response obejct as a
...
+#### Handles journey forking
+
+Each step definition accepts a `next` property, the value of which is the next route in the journey. By default, when the form is successfully submitted, the next steps will load. However, there are times when it is necessary to fork from the current journey based on a users response to certain questions in a form. For such circumstances there exists the `forks` property.
+
+In this example, when the submits the form, if the field called 'example-radio' has the value 'superman', the page at '/fork-page' will load, otherwise '/next-page' will be loaded.
+
+```js
+
+'/my-page': {
+  next: '/next-page',
```
Comment (gavboulton) [inline README.md]: Could you indent wit 4 spaces please (it'll match the rest of the readme).

File: test/spec/spec.form.js [hunk 13]
```diff
@@ -515,14 +514,136 @@ describe('Form Controller', function () {
...
+            });
+
+        });
+
+    });
+
+    describe('successHandler', function () {
+
+        beforeEach(function () {
+            Form.prototype.getNextStep = sinon.stub();
```
Comment (JoeChapman) [inline test/spec/spec.form.js]: Are you tearing this down afterwards?

Comment (joefitter) [inline test/spec/spec.form.js]: I am now 👍

**Comment by joefitter** [comment]
@JoeChapman @gavboulton amends made

**Comment by JoeChapman** [comment]
Ok, LGTM. @gavboulton can we push the button?

**Comment by joefitter** [comment]
tests failing - 2 secs

**Comment by joefitter** [comment]
G2G @gavboulton @JoeChapman

**Comment by joefitter** [comment]
To note - this PR moves the current functionality from HOF to hmpo-form-controllers - there are issues at a middleware level which are fixed in this PR - https://github.com/UKHomeOffice/passports-form-wizard/pull/52 - if this is merged then the middleware fixes (backLinks and prereqs) should be merged too

**Comment by gavboulton** [comment]
Thanks @joefitter, hmpo-form-controller@0.7.0 published.
